### PR TITLE
Rewrite arch/linux

### LINKED ISSUE #4201: IPv4 routing table is missing entries for additional addresses on the loopback interface
### Brief description

Scapy IPv4 routing table is missing additional addresses configured on the `lo` interface.

### Scapy version

2.5.0

### Python version

3.11.2

### Operating system

Debian bookworm 12.2 running kernel 6.1.52-1

### Additional environment information

_No response_

### How to reproduce

 Given the following IPv4 and IPv6 addresses assigned to `lo`:
```
vgutierrez@ncredir4001:~$ ip addr show dev lo
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet 198.35.26.98/32 scope global lo:LVS
       valid_lft forever preferred_lft forever
    inet6 2620:0:863:ed1a::3/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute 
       valid_lft forever preferred_lft forever
```

scapy provides an IPv6 routing table that includes `2620:0:863:ed1a::3/128`:
```
>>> conf.route6
Destination                     Next Hop  Iface   Src candidates              Metric
2620:0:863:101::/64             ::        ens13   2620:0:863:101:10:128:0:32  256   
fe80::/64                       ::        ens13   fe80::a800:ff:fe39:c6fc     256   
fe80::/64                       ::        ipip60  fe80::cbd:4aff:fe6b:a7c5    256   
::1/128                         ::        lo      ::1                         0     
2620:0:863:101:10:128:0:32/128  ::        ens13   2620:0:863:101:10:128:0:32  0     
2620:0:863:ed1a::3/128          ::        lo      ::1                         0     
fe80::cbd:4aff:fe6b:a7c5/128    ::        ipip60  fe80::cbd:4aff:fe6b:a7c5    0     
fe80::a800:ff:fe39:c6fc/128     ::        ens13   fe80::a800:ff:fe39:c6fc     0     
::/0                            fe80::1   ens13   2620:0:863:101:10:128:0:32  1024 
```

But the IPv4 route table is missing `198.35.26.98/32`:
```
>>> conf.route
Network     Netmask        Gateway     Iface  Output IP    Metric
0.0.0.0     0.0.0.0        10.128.0.1  ens13  10.128.0.32  0     
10.128.0.0  255.255.255.0  0.0.0.0     ens13  10.128.0.32  0     
127.0.0.0   255.0.0.0      0.0.0.0     lo     127.0.0.1    1
```

In fact, if we compare scapy IPv4 route table against the kernel local routing table, quite some entries seem to be missing:
```
$ ip route show table local
local 10.128.0.32 dev ens13 proto kernel scope host src 10.128.0.32 
broadcast 10.128.0.255 dev ens13 proto kernel scope link src 10.128.0.32 
local 127.0.0.0/8 dev lo proto kernel scope host src 127.0.0.1 
local 127.0.0.1 dev lo proto kernel scope host src 127.0.0.1 
local 127.0.0.42 dev ipip0 proto kernel scope host src 127.0.0.42 
broadcast 127.255.255.255 dev lo proto kernel scope link src 127.0.0.1 
local 198.35.26.98 dev lo proto kernel scope host src 198.35.26.98 
```

### Actual result

_No response_

### Expected result

_No response_

### Related resources

_No response_

### PR DESCRIPTION
**Pull request by gpotter2**
This rewrites most of the arch/linux code that handles the loading of `conf.ifaces`, `conf.route` and `conf.route6`, in order to use a RTNETLINK socket (instead of the current reading of `/proc/net/XXX`).

Among those:
- the read_routes(6) functions
- the linux interfaces provider
- arch/linux util functions

This:
- adds support for multiple IPv4 addresses to interfaces
- handles interfaces aliases, etc. natively without needing additional tricks
- allows us to have filter routes by routing tables, also giving us access to 'local', rather than only all tables, giving us generally much better handling of routes.

fixes #4201
closes #3275
(maybe other issues too?)

## Note to all maintainers

This has been tested on 32 big endian architecture thanks to the folks at packit. (woooo)

This might need thorough **TESTING** on other platforms (that we don't include in the CI). I hope I didn't break anything on *BSD.

### CODE REVIEW

**Comment by gpotter2** [comment]
@guedou @p-l- This is ready for review. If you could just try playing around with routes, it would help a lot.

Just note an intended change: for IPv4, 'via' and 'default' routes show '0.0.0.0' as the source, but that gets resolved properly when calling `conf.route.route`.

@evverx Hi ! I was wondering, if you have the time, if it was possible to test this PR with your `packit` setup. (BTW, If you have the time, feel free to submit a PR to include it here. It looks like It's working great). I'm a bit worried that this PR breaks on some plateforms, because of the rnetlink implementation. Especially on big-endian or 32-bits platforms. Thanks a lot.

**Comment by evverx** [comment]
@gpotter2 I cherry-picked the commit on top of my fork and triggered Packit. Looks like the tests passed on all the architectures: https://copr.fedorainfracloud.org/coprs/packit/evverx-scapy-2/build/7329726/. I'll open a PR a bit later.

**Comment by evverx** [comment]
Looks like with this PR applied I can see the following log messages in the journal when the test suite is run
```
Apr 20 03:10:26 H kernel: __nla_validate_parse: 6 callbacks suppressed
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.                                                             
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.                                                             
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.                                                             
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.                                                             
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.                                                             
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.                                                             
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.                                                             
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.                                                             
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.                                                             
Apr 20 03:10:50 H kernel: netlink: 4 bytes leftover after parsing attributes in process `python'.        
```
I switched back to the master branch, ran the tests and the logs disappeared.

**Comment by gpotter2** [comment]
Thanks a lot. This should be fixes in the latest version.

**Comment by evverx** [comment]
Looks like the kernel no longer complains.

> fixes https://github.com/secdev/scapy/issues/4201
> (maybe other issues too?)

As far as I understand it should also address https://github.com/secdev/scapy/issues/2133 and https://github.com/secdev/scapy/pull/3275.

**Review by guedou — APPROVED** [review]
Awesome PR! Thanks a lot for made it happen. I bet that it will also speed up loading Scapy on systems with many routes.

What's the minimal kernel version that is needed for this PR to work?

**Comment by gpotter2** [comment]
So just a heads up: I'm really not sure this speeds up loading. It gives a much more more accurate result on my machine and in several test setups I tried, but I haven't tested it on a machine that has a billion routes, or similar. I'm not really sure how it performs on those, so maybe it would be a good idea to try that out.

File: scapy/arch/linux/rtnetlink.py [hunk 1]
```diff
@@ -0,0 +1,851 @@
...
+        rtmsghdr(
+            nlmsg_type="RTM_GETADDR",
+            nlmsg_flags="NLM_F_REQUEST+NLM_F_ROOT+NLM_F_MATCH",
+            nlmsg_seq=1,
+            nlmsg_pid=pid,
+        )
+        / ifaddrmsg(
+            ifa_family=af_family,
+            data=[
+                ifaddrmsg_rtattr(),
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/arch/linux/rtnetlink.py:676]: We don't. fixed

**Comment by guedou** [comment]
I tried to load 1000 routes using the following shell and Python scripts, and got the following error:

```shell
bash bench_load_prefixes.sh 
gpotter2/rewrite-linux-arch 1000
9

real	0m1.848s
user	0m0.766s
sys	0m0.389s
root@vagrant:/vagrant# vim bench_load_prefixes.sh 
root@vagrant:/vagrant# bash bench_load_prefixes.sh 
gpotter2/rewrite-linux-arch 10000
Traceback (most recent call last):
  File "/vagrant/test_routes_loading.py", line 1, in <module>
    from scapy.all import *
  File "/vagrant/scapy/all.py", line 26, in <module>
    from scapy.route import *
  File "/vagrant/scapy/route.py", line 222, in <module>
    conf.route = Route()
                 ^^^^^^^
  File "/vagrant/scapy/route.py", line 37, in __init__
    self.resync()
  File "/vagrant/scapy/route.py", line 47, in resync
    self.routes = read_routes()
                  ^^^^^^^^^^^^^
  File "/vagrant/scapy/arch/linux/rtnetlink.py", line 796, in read_routes
    results = _read_routes(socket.AF_INET)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vagrant/scapy/arch/linux/rtnetlink.py", line 772, in _read_routes
    results = _sr1_rtrequest(
              ^^^^^^^^^^^^^^^
  File "/vagrant/scapy/arch/linux/rtnetlink.py", line 647, in _sr1_rtrequest
    msgs = rtmsghdrs(sock.recv(65535))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vagrant/scapy/base_classes.py", line 398, in __call__
    i.__init__(*args, **kargs)
  File "/vagrant/scapy/packet.py", line 183, in __init__
    self.dissect(_pkt)
  File "/vagrant/scapy/packet.py", line 1076, in dissect
    s = self.do_dissect(s)
        ^^^^^^^^^^^^^^^^^^
  File "/vagrant/scapy/packet.py", line 1014, in do_dissect
    s, fval = f.getfield(self, s)
              ^^^^^^^^^^^^^^^^^^^
  File "/vagrant/scapy/fields.py", line 1843, in getfield
    raise MaximumItemsCount(
scapy.fields.MaximumItemsCount: Maximum amount of items reached in PacketListField: 100 (defaults to conf.max_list_count)

real	0m1.329s
user	0m0.542s
sys	0m0.260s
```

Setting the value of `conf.max_list_count` to 1000 in `scapy/config.py` fixed this issue, but it seems a bit too much to change this value globally only to parse routes.

You can put random IPv4 prefixes into prefixes.txt.gz or populate it from real ones (as I did).

```shell
cat bench_load_prefixes.sh 
N=1000
for PREFIX in $(zcat prefixes.txt.gz |head -n $N)
do
  sudo ip route add $PREFIX dev eth0
done

echo $(git branch --show-current) $N
time python3 test_routes_loading.py

for PREFIX in $(zcat prefixes.txt.gz |head -n $N)
do
  sudo ip route del $PREFIX dev eth0
done
```

```python
from scapy.all import *
print(len(conf.route.routes))
```

**Comment by guedou** [comment]
Empirical benchmarks show similar results (i.e. 2s to 5s loading time) between master and this PR. However, for approximately 50000 routes, master loads in 7s while this PR loads in 41s. I was not expecting this PR to be slower when dealing with more routes.

I did the test on Ubuntu 23.10 with kernel 6.5.0-15-generic.

**Comment by gpotter2** [comment]
I expected that, we're popping a socket and parsing several messages instead of simply opening a file descriptor and parsing some text.

That being said, it's the only way of solving the 'how do you not load the BGP table' problem.

**Comment by gpotter2** [comment]
> I tried to load 1000 routes using the following shell and Python scripts, and got the following error:

This is fixed.

**Pull request merged**